text/0000-github-access-policy.md

```diff
@@ -0,0 +1,58 @@
+- Feature Name: rust-lang_github_org_access_policy 
+- Start Date: 2020-03-02 
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a policy for managing permissions to the [Rust-Lang GitHub Organization](https://www.github.com/rust-lang) and repositories within this organization.
+
+This RFC was written in consultation with the Governance Working Group and the Infrastructure team. Most discussion took place on [this issue](https://github.com/rust-lang/wg-governance) and [this pull request](https://github.com/rust-lang/wg-governance/pull/42).
+
+# Motivation
+[motivation]: #motivation
+
+Access control for the [Rust-Lang GitHub Organization](https://www.github.com/rust-lang) and repositories within that organization is currently managed ad-hoc. We need a policy that defines how these accesses are granted and managed. This will allow us to have greater security in permissions to our GitHub org and also allow the infra team to build appropriate tooling to automate access control when possible.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Rust-Lang GitHub Permissions Policy
+This policy applies to both the [Rust-Lang GitHub Organization](https://github.com/rust-lang/) and all repositories within that organization.
+
+### Rust-Lang Organization
+Membership in the Rust-Lang GitHub organization is managed by the organization owners.
+
+All members of the [Core Team](https://github.com/rust-lang/team/blob/master/teams/core.toml) have the right to be organization owners.
+
+Selected members of the [Infrastructure Team](https://github.com/rust-lang/team/blob/master/teams/infra.toml) can also be organization owners if their work requires it.
+
+Owners should use a separate account from their main GitHub account dedicated to managing the organization. This account may not be used to commit code and must have 2FA enabled.
```

Comment:
I personally believe this isn't a productive suggestion. While the org-owner accounts do have very extensive access, and we should push for users to have good security behavior more generally, I don't think a separate account helps us much. In practice, if there's a bug in GitHub's access controls (e.g. an XSS attack or similar) then even a "normal" account for most Rust team members is still pretty bad; that account likely has sufficient privileges to be able to, for example, delete every issue on the rust-lang/rust repository -- moderation-level access is necessary for daily activities (editing descriptions on issues/PRs) within that repository.

If we're concerned about the potential threat vector of "stolen device", then we should instead provide easy and readily available access points for a member to self-remove their access. (Currently it could take several hours for e.g. a team repo PR to land and take effect).

I generally feel that restricting access to our repositories isn't necessarily worthwhile; in practice we need to trust that people are doing due diligence in terms of security, both within the browser (e.g. GitHub) and separately, i.e. code review, etc.

I feel that generally speaking if we're interested in restricting this access, we need to either push for GitHub to implement better access controls (for example, only editing issue descriptions, nothing more) or migrate away from GitHub.

Reply:
The challenge here is that it is extremely difficult for a user to ensure their GitHub account hasn't been compromised if they use any apps (especially OAuth apps). (See the [detailed discussion](https://github.com/mozilla-services/GitHub-Audit/blob/master/docs/threat.md) if interested.)

It is also impossible for an external audit process to detect account compromise. Only the account holder has access to much of the data. (Good privacy!)

At Mozilla, our experience has been that owner level privileges are hardly ever needed. That reduces the burden of owners using two accounts to the point where it is a reasonable "insurance policy" we are willing to pay for. (Across 20 some orgs, and thousands of members, I only have to use owner permissions 2-3 times a week.)

Reply:
Hm, so I agree with you that I don't use owner-level permissions essentially ever, but owner permissions on an organization do mean that *any* repository within rust-lang automatically grants core team members (the current owners) permissions for moderation and otherwise editing.

I would be unopposed, I think, if we granted that permission via some automation that granted admin (or at least write) access to repositories within the org. I would also be okay with that having to be enabled on a case by case basis (e.g., I would ask a bot somewhere for access, and that would be enabled temporarily).

Reply:
Hello @Mark-Simulacrum!

I think I understand your concerns, but want to make sure.

Let's say person A has Account 1 (without owner level permissions - the one they should be using for their daily GitHub user) and Account 2 (without owner level permissions - only used when absolutely needed).

As Account 2 will not have owner level permissions, it will not automatically have permissions for moderation or editing. Those will need to be enabled through the teams repo (like with any other person who is not an owner), which can take a few hours for someone to put in a PR, get it reviewed, then merge it.

Parsing the concern that way - I'm honestly ok with person A's Account 2 needing to go through the review process on the teams repo. If there is ever an emergency - something needs to be moderated immediately - they can always switch to their owner's account (Account 1) to do so. It's a bit of inconvenience - but it buys us a lot, and ensures that someone only uses the account when they absolutely need those privileges.

Does that address your concerns?

Reply:
Yes (we also spoke separately). I'm not entirely satisfied, but I think the risk vector of OAuth apps and such has convinced me sufficiently that I can withdraw my concerns here.